### ISSUE TITLE
PPSSPP menu disappears on recent builds.

### ISSUE THREAD

**Issue by Rekrullurker** [hrydgard/ppsspp#9647]
This is with the 32-bit Windows version under XP Pro, SP3.

I'll have to download a bunch of versions to test if you need to know exactly when this started, but it's been happening sporadically for at least a couple weeks now. It seems to occur more frequently with some versions than with others.

I run PPSSPP, the main screen comes up, I click on Settings and when I return to the main screen, often parts of the onscreen menu will be missing. It might be the drive/games list, it might be the "Recent, Games, Homebrews & Demos" at the top, it might be the options on the right, some combination of these, or they might all be missing and the screen will just be blank.

Sometimes this happens if I just switch between "Recent" and "Games" or when switching drives. It happens regardless of whether I click the back button from settings or whether I press ESC.

Basically any time I switch screens in the emulator, there's a chance that some or all of the GUI on the main screen will be missing.

The settings and credits screens never have this problem, it's only the main screen that loses parts.

If I quit the emulator and restart it, everything is back to normal, but it will usually happen again after switching screens.

**Comment by xiushudongfang** [comment]
Yeah,the same bug happned on win10.

**Comment by hrydgard** [comment]
Sounds odd. Any connection in particular with using the OpenGL, D3D9 or D3D11 backend?

**Comment by Rekrullurker** [comment]
I usually use OpenGL, but I just tried it with D3D9 and got the same results. I can't use Vulkan apparently.

My GPU is an NVIDIA GeForce GT 430 with the latest available drivers that support this card; 368.81.

**Comment by unknownbrackets** [comment]
If you go back to 1.3.x, does this still happen?  Or does it start happening as of some specific git build?

-[Unknown]

**Comment by Rekrullurker** [comment]
I just tested a bunch of builds and it seems to have started with 1.4-52. At least I couldn't make it happen in the earlier builds. I ran them several times each, switching screens and they all seemed fine. When I try that in 1.4-52, it usually takes less than a minute of playing around with it before parts of the menu disappear.

Note that it doesn't happen all the time. One time I might run it and everything works as it should. Then I quit it, run it again and maybe this time it messes up after switching screens 2-3 times. Sometimes it happens on the first try and sometimes it takes several tries before it happens.

On 1.4-142 it happens almost every single time when I switch screens.

**Comment by hrydgard** [comment]
Can you take a screenshot of what it looks like when it's broken and post here?  (just drag drop to post images). That might help us figure out in what way it's broken.

I can't reproduce it on any of my machines but they're both much more recent nVidia.

**Comment by Rekrullurker** [comment]
Here are four screenshots. I couldn't get everything to disappear like it was doing the other day, but I'm sure you can imagine what that would look like.

Like I said, it doesn't happen all the time. Sometimes it works exactly as it should and other times, I come back from the settings and parts of the main screen are missing. It seems to be random (at least I can't figure out a pattern to it) what parts will disappear.

Once something has disappeared, it never seems to come back no matter how many times I switch screens. I have to quit PPSSPP and run it again to get everything back to normal.

![capture_004_28042017_112047_687](https://cloud.githubusercontent.com/assets/23264794/25535540/76313950-2c05-11e7-9b3b-3d17b8328cce.png)
![capture_001_28042017_111851_828](https://cloud.githubusercontent.com/assets/23264794/25535543/763b1f4c-2c05-11e7-8dd7-6aa45834007d.png)
![capture_002_28042017_111927_078](https://cloud.githubusercontent.com/assets/23264794/25535541/7634a2a2-2c05-11e7-8fa6-472b61be8cf2.png)
![capture_003_28042017_112034_109](https://cloud.githubusercontent.com/assets/23264794/25535542/7638e308-2c05-11e7-8a2d-a7703466599d.png)

**Comment by unknownbrackets** [comment]
So, you can't make this happen in v1.4-51-ga8d6de2, but you can make it happen in v1.4-52-g19bf222 - right?  That change (19bf222ea423ab8f394c6f9219e84b87a50fde03) doesn't seem like it could cause this problem.

I wonder - maybe this is somehow related to the window size.  Does it help to use windowed mode with 1x window size?  Or does it still happen?

Also, are you using a non-standard DPI?
http://www.desksite.com/faq/dpi-settings/windowsxp.htm

96 is standard DPI.  Maybe this only happens when using a higher DPI (which is a bug if so - just trying to narrow down the cause.)

-[Unknown]

**Comment by Rekrullurker** [comment]
So far I haven't been able to make it happen at all in 1.4-51. I just spent about 15-20 minutes running that version, switching screens, etc, and it worked perfectly every time. In 1.4-52, it seems to happen at least half the time.

It happens in both windowed mode and fullscreen with the resolution set to 1x (I normally leave it set to 1:1 Auto, but I did try changing it to 1x just to be sure).

My DPI is set to the standard 96. I also tried changing my desktop resolution (I normally use 1024x768, but I also tried 1280x1024) and that didn't make any difference.

I made a video of the bug. I don't have screen recording software, so I used a camera. I apologize for the shaky camera work, I don't have a tripod and it's a little hard to aim the camera and see what I'm doing. :)

I can't upload the video here, so I put it on Sendspace. If there's a better hosting site to use, please let me know. The video is 23MB;

https://www.sendspace.com/file/2o1dnf

You'll notice that sometimes it works as it should and sometimes it just randomly deletes part of the main screen.

**Comment by hrydgard** [comment]
I wonder if it may be that one after all - the views that disappear are all scrollable, and the scroll pos might be corrupted somehow so they're scrolled off into oblivion? maybe a NaN sneaks in somewhere? I'll look at it.

**Comment by unknownbrackets** [comment]
Ah... maybe `IsGestureActive` is being called while `scrollTouchId_` is -1, then?  Didn't think of that.

-[Unknown]

**Comment by hrydgard** [comment]
I added some extra checks but yeah, I don't see it...

**Comment by hrydgard** [comment]
When parts are missing, are they still clickable and thus just invisible, or do they not react?

**Comment by Rekrullurker** [comment]
No. When something disappears, it can't be clicked on. I just tried it multiple times and if I couldn't see it, I couldn't click on it.

BTW, I just tried 1.4-165 and got the same behavior.

**Comment by hrydgard** [comment]
Thanks, that at least removes some possibilities. I'm still baffled by this, can't get it to happen...

**Comment by unknownbrackets** [comment]
Just to be sure, I've tried the latest git on a 32-bit Windows XP SP3 VM (this isn't using the NVIDIA GPU driver, though), with font smoothing disabled, just to be sure.  I feel like there must be something about this system or its configuration causing the issue.

What processor do you have?  Not sure if it could matter, but we might have some SSE4 path somewhere.

If you change the filename "PPSSPPWindows.exe" to something else, like "My PPSSPP Program.exe" (not the shortcut, but the actual exe, like in the extracted zip file), does anything change?  Sometimes other programs hook into programs by filename.

Go to Properties on the shortcut, then Compatibility, make sure none of the checkboxes are filled.

-[Unknown]

**Comment by Rekrullurker** [comment]
I have a 2.4Ghz Core-2-Duo processor.

I just tried the latest build; 1.4-167 and the bug still happens.

No compatibility options are checked and renaming the EXE file didn't make a difference.

**Comment by unknownbrackets** [comment]
Okay, so you definitely don't have SSE 4.2 and you *may* not have SSE 4.1.  But I can't find anything that uses SSE4 in the paths that render the UI...

-[Unknown]

**Comment by Rekrullurker** [comment]
No, I don't have SSE 4.1;

![capture_001_02052017_180745_953](https://cloud.githubusercontent.com/assets/23264794/25641577/d7d52d6e-2f62-11e7-9e14-13c30d766fe1.png)

**Comment by LunaMoo** [comment]
Maybe this isn't about v1.4-52-g19bf222 changes, but the build itself?

 Comparing the size of v1.4-51-ga8d6de2 and v1.4-52-g19bf222 amd64 binaries are same, however with x86 the newer commit is slightly smaller. Was the buildbot updated around that time by chance? Don't have older one anymore, but I did had both for a while and VS2017 was for sure producing slightly smaller binaries than 2015 one.

**Comment by pal1000** [comment]
I can replicate it too, but only for 32-bit builds downloaded from Orphis buildbot, Self made 64-bit builds seams to be unaffected. I never did 32-bit builds though. I upgraded to Visual Studio 2017 quickly after release as it is more clean than Visual Studio 2015 which was horribly bloated for my taste and took advantage of Creators Update timing to ditch with all potential remnants. Visual Studio 2017 [had a bunch of updates in quick succession until recently](https://www.visualstudio.com/en-us/news/releasenotes/vs2017-relnotes). If Orphis buildbot did update, it may not be on latest patch level v15.0.26403.7 with C/C++ toolset v14.10.25017. If I remember correctly the toolset was updated 2 times since release.

**Comment by hrydgard** [comment]
@Orphis , any change you did to the buildbot at that time?

**Comment by unknownbrackets** [comment]
You're right - this easily reproduces with the buildbot 32-bit build.  I stupidly tested my own compile only.

-[Unknown]

**Comment by hrydgard** [comment]
Huh. I just downloaded the latest 32-bit from Orphis' buildbot and nope, I still can't repro. Started a few games, clicked around a lot, into menus and out of them, all works. Tried both mouse and keyboard navigation.

By the way, to those who can repro, does it repro if you only move around in the emulator with the keyboard?

**Comment by hrydgard** [comment]
I got a single repro twice, only homebrew list disappeared. Could get it back after resizing the window a few times, funnily enough. And only with @Orphis builds...  so weird.

**Comment by Rekrullurker** [comment]
I just tested 1.4-167 (32-bit) and the bug only seems to happen when I use the mouse. If I use the keyboard or a gamepad, I can't seem to make it happen. I went in and out of the settings and menus and nothing disappeared.

Also, I've just discovered something;

I don't have to actually click on any of the UI elements to make things disappear, just clicking anywhere on the emulator screen/in the window is enough to make some of the text instantly vanish. So it seems that it's the mouse click that is causing it, the screen switching was just masking that fact.

I made another video;

https://www.sendspace.com/file/dv6atx

**Comment by hrydgard** [comment]
Thanks for your testing! This gets weirder and weirder! But yeah, I think we've narrowed it down to being mouse-related. The buildbot-only thing could be an old compiler on the buildbot, either with a bug, or there's a real bug but it just happens to have worse consequences with that compiler.

**Comment by hrydgard** [comment]
Added two experimental commits. The result will give us information whether they resolve the issue or not.

**Comment by unknownbrackets** [comment]
Those didn't help.

The view that disappears has a y of -nan.  `scrollPos_` is also -nan, but `scrollTarget_` is 0.  `inertia_` is -nan.  I set a breakpoint here:

```c++
		if (gesture_.GetGestureInfo(gesture, input.id, info)) {
			inertia_ = info[1];
		}
*		scrollTouchId_ = -1;
```

For `inertia_ != (inertia_ * 1.0f)` and it tripped.  `input.id` is 0, `gesture` is `GESTURE_DRAG_VERTICAL`, and the info is `-nan, -nan, 0, 0`.  Looking at `gesture_`, ALL pointers are filled ff (including `down` and `active`.)  There's nothing wrong with the actual TouchInput event itself.

It looks as if `gesture_` is somehow not being zero-initialized.

(I got this far by making the buildbot generate a pdb.)

-[Unknown]

**Comment by hrydgard** [comment]
Scary indeed. @Orphis , can you upgrade the compiler on the buildbot to the latest?

Also I might memset all arrays that we auto-initialize that way...

**Comment by hrydgard** [comment]
It might be the fact that GestureDetector is lacking a defined constructor that is confusing the compiler. I'm gonna make a separate commit that tests that, and then memset in the next.

**Comment by hrydgard** [comment]
Those two commits done, though forgot to link this issue this time.

**Comment by Rekrullurker** [comment]
It seems to be fixed for me as well. I just tested 1.4-171 and can no longer make anything disappear. :)

**Comment by hrydgard** [comment]
Yeah, unknown's digging found it. I worked around the compiler bug. 

Can you guys also test 1.4-170? It's interesting if that one works or still breaks.

**Comment by unknownbrackets** [comment]
I can still reproduce with v1.4-170-gfb95288, so having a trivial constructor wasn't enough to workaround the compiler bug.

-[Unknown]

**Comment by Rekrullurker** [comment]
Same for me, I still get the bug with 1.4-170.

**Comment by pal1000** [comment]
Here too. I was able to repro with v1.4-170-g fb95288.

**Comment by hrydgard** [comment]
Thanks guys. I'm closing this, still worried about other potential bugs though..
